### ISSUE TITLE
Taking a step back

### ISSUE THREAD

**Issue by westnordost** [ENT8R/tags2names#1]
Answer to https://github.com/westnordost/StreetComplete/issues/309#issuecomment-356354053

> What do you think about it?

I think it is very bold of you to tackle this problem and create a reusable library for that. In other words, it's awesome! Awesome enough, to get me worrying about it very much that it should be done right. Here is what I mean:

If this can be done, this library would be worth gold for the OpenStreetMap project. Literally every editor, geocoder, map renderer, routing app etc., would profit from a commonly maintained database of names and aliases as now, everybody is brewing his own soup. This is hugely inefficient, and if this project can become an easy plug-in alternative to their own solutions, this will be hugely popular.

But this requires a well design of how to structure the translation data, taking into account all the different use cases for such a dictionary of OSM tags to names.
Most users would just like to have a name, others require names in different grammatical cases (like SC). Others yet, I am thinking of geocoders, would like to find the tag (combination), fitting to a name or alias, so the other way round.

In any case, you probably know this, the code-part of that is not the biggest part, **but the data**. The success of this library stands and falls with the availability and maintainability of the translations. So, the data (the translations) should be arranged in a format, that can be im- and exported from a crowd sourcing translation platform (like POEditor and alternatives) easily.

I would like you to step back a bit and review what you have done so far. The better this data structure is designed, the more useful it will be. You could of course say, nah, it's fine, and just focus on this one use case, but then it will be, yet again, only a home-brew soup and I think, as said, it would miss a huge potential.
All in all, it doesn't have to be that complex, I am just saying that diving into trying to solve [one issue](https://github.com/westnordost/StreetComplete/issues/309), look at the big picture. Perhaps, write down the use cases that should be covered, and a list of requirements for the structure, then implement it. 

--- 

So, already started such a project in 2015, but at some point decided that I should pause this for now and first start with the app (StreetComplete), then later come back to it. That did not happen, because I have all hands full with this app.

For reference, I uploaded it to github just now: https://github.com/westnordost/tagdictionary

I think it should work. I did not test it now.

I already put a great deal of thought into it, there should be a file that explains my considerations. But, that project had some unsolved issues as well. Namely:
- no thought given about different grammatical cases
- data structure is (probably) not fit to stick it into one of those crowdsourced translation platforms. I think this is actually pretty important. Translations are one of the things that can be crowdsourced excellently

So, what you do with that, is up to you. For now, just consider it as a reference. But if you see potential and think the mentioned conceptual problems can be accounted for, it might also be a consideration to take that as a base.

**Comment by ENT8R** [comment]
> So, I already started such a project in 2015

Oh cool, I didn't know that you already worked on such a project...

I think too that this project can have a big potential if it is done right. However this should be _only_ a proof of concept right now and I already thought about how to do the translation (which will be pretty hard, because nearly every country has its own rules...).
Sure, I have to put more thoughts into all the possible use cases and how the data has to be formatted and structured...
This dictionary is, right now, pretty useless, because there is only support for English and German. But this can only be changed, if translators can translate it into different languages. So, I will follow your proposal and take a step back 👍 to think about the translation problem, possible use cases and all the other things which are important!
Thank you for giving me food for thought, I will try to follow most or even all of your proposals, as they make all sense to me!

**Comment by westnordost** [comment]
Keep me up to date, perhaps I can help. I have been yearning to get this thing started again - and be it that I contribute to your repository for a change. :-P

**Comment by rugk** [comment]
Thanks @westnordost for thinking about the whole FLOSS/OSM community,that's really great!

BTW as for translation tools/libs I've also discovered some new things from Mozilla lately: https://pontoon.mozilla.org/ and http://projectfluent.org/
The former may not help here as this is no website, but the later looks interesting from having a quick look at it only.

**Comment by callis** [comment]
Wikidata might be another option!?
https://www.openstreetmap.org/user/PlaneMad/diary/43105

**Comment by ENT8R** [comment]
> Wikidata might be another option!?

Well, this is basically an existing dictionary, isn't it? With this project, a general OSM tag dictionary is almost useless...

> BTW as for translation tools/libs I've also discovered some new things from Mozilla lately: https://pontoon.mozilla.org/ and http://projectfluent.org/

For me Pontoon looks like this is only a translation tool for the web, or am I wrong?
But Project Fluent looks quite promising as it seems to have the ability to translate ICU's MessageFormat

**Comment by ENT8R** [comment]
Does somebody have experience with the data of wikidata? From a quick look it looks very promising to me... So we could use the data from wikidata and then we just need additional country-wise data to deal 
 for example with the genus of a word

**Comment by callis** [comment]
Maybe wikidata is even willing to adapt or complement their database to the needs of OSM. Especially as there is already some collaboration between the two projects.
https://www.wikidata.org/wiki/Wikidata:OpenStreetMap
https://wiki.openstreetmap.org/wiki/Wikidata

**Comment by ENT8R** [comment]
I just imported the plain values from WikiData... I will think about the next step: Whether those strings should be translated by the community or trust the values of WikiData and import also the values for other languages. But I tend to use the first solution as this helps to review the quality of the values and whether the data can be used at all...

**Comment by rugk** [comment]
Can't you use a random sample to check the quality?

**Comment by ENT8R** [comment]
> Can't you use a random sample to check the quality?

Yes, I already did that but I could not really find an issue... For me it all looks good. The only mistake I spot is that sometimes words which should begin with a lowercase letter begin with a capital letter...

**Comment by matkoniecz** [comment]
> Wikidata might be another option!?
> https://www.openstreetmap.org/user/PlaneMad/diary/43105

Wikidata is not a good solution for reasons mentioned in https://www.openstreetmap.org/user/PlaneMad/diary/43105

> It is known that OSM tags don't always map 1:1 exactly with existing Wikidata concepts


> Wikidata is flexible enough to allow anyone to create a new concept if it does not overlap with an existing one, so its fairly easy to correct the definition in such situations to align almost perfectly with the OSM concept.


> Agreed that we could create wikidata entries for every "OSM tag XYZ" as distinct from "the English concept XYZ"; I'm just not clear what that buys us over and above an existing OSM wiki article that can be translated into other languages? It just makes the whole thing harder to maintain.

It would just be copying OSM wiki into different platform making editing significantly more complicated, exposed to harmful/useless editing by non-OSM users for no benefit.

> Can't you use a random sample to check the quality?

I looked at https://query.wikidata.org/embed.html#%0ASELECT%20%3Fosm%20%3Fwd%20%3FwdLabel%20%3FwdDescription%20%3Fsitelink%20WHERE%20%7B%0A%20%20%23%20Find%20all%20Wikidata%20items%20with%20an%20OSM%20tag%0A%20%20%3Fwd%20wdt%3AP1282%20%3Fosm%20.%0A%20%20%23%20Look%20for%20sitelinks%20in%20the%20needed%20language%0A%20%20OPTIONAL%20%7B%20%3Fsitelink%20schema%3Aabout%20%3Fwd%20.%20%3Fsitelink%20schema%3AinLanguage%20%22en%22%20%7D%0A%20%20%23%20Set%20the%20label%20language%0A%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22en%22%20%7D%0A%7D%0AGROUP%20BY%20%3Fosm%20%3Fwd%20%3FwdLabel%20%3FwdDescription%20%3Fsitelink linked in the diary entry - and many descriptions are useless for explaining OSM tags (even where relation is correct), just from a quick look I found mistagged relations (that I removed from OSM wiki and Wikidata).

**Comment by ENT8R** [comment]
> Wikidata is not a good solution for reasons mentioned in https://www.openstreetmap.org/user/PlaneMad/diary/43105

Well, I thought they said that not all descriptions are good but this project does not even use the descriptions but only the names...

> It would just be copying OSM wiki into different platform making editing significantly more complicated, exposed to harmful/useless editing by non-OSM users for no benefit

Mmhh... That's true! Having more than 1200 different values to maintain is not that funny :)

---

So how do we want to continue? Maybe use taginfo as a source as said in openstreetmap/iD#2708 or add all values by hand? I would prefer to use the first solution :)

**Comment by matkoniecz** [comment]
@ENT8R As I understand it the hard part is deciding what kind of data is necessary - primarily, what about different grammatical cases?

**Comment by ENT8R** [comment]
There is no source I found so far which takes care of grammatical cases... So this has to be added by hand. Another thing which is difficult with that is that e.g. in English there is nothing like a genus like e.g. in German. So it is nearly impossible to translate from English to German...

**Comment by matkoniecz** [comment]
> So this has to be added by hand

Yes, I think I would separate preparing data format from translating itself that would happen on Transifex/poeditor/similar platform.

**Comment by matkoniecz** [comment]
My main motivation is to create part that is necessary for StreetComplete.

> If this can be done, this library would be worth gold for the OpenStreetMap project. Literally every editor, geocoder, map renderer, routing app etc., would profit from a commonly maintained database of names and aliases as now, everybody is brewing his own soup. This is hugely inefficient, and if this project can become an easy plug-in alternative to their own solutions, this will be hugely popular.

If that should be useful for others I would start from asking what kind of data and in which form would be useful. I would probably start from StreetComplete, prepare very short description of what this project is trying to do and ask developers of other projects.

I wonder about the best way to do that. Open issues on bug trackers? Directly mail the most prominent coders? Ask on a dev mailing list? Ask on project specific mailing lists if one exists? Do all of that?

**Comment by rugk** [comment]
Could not Wikidata or so help? We would have a common/major free source/platform.

**Comment by rugk** [comment]
As for your last question IMHO it's good to handle it publicly, it's FLOSS after all😊. So I would suggest issues.

**Comment by matkoniecz** [comment]
> Could not Wikidata or so help? We would have a common/major free source/platform.

Searching "Wikidata" on this page would found https://github.com/ENT8R/tags2names/issues/1#issuecomment-370125753 among others

**Comment by rugk** [comment]
As it is all about tags, maybe integrating it into _taginfo_ somehow may be the best? (with it's JSON data loading already implemented there for tag usage, at least)
